File: Android/src/androidMain/kotlin/com/taskbridge/android/repository/impl/NavigationRepositoryImpl.kt
package com.taskbridge.android.repository.impl

import com.taskbridge.android.repository.NavigationRepository
import com.taskbridge.core.interactors.navigation.NavigationInteractor
import com.taskbridge.core.models.navigation.AppTab
import com.taskbridge.core.models.navigation.NavigationDestination
import com.taskbridge.core.models.navigation.NavigationDestinationMessage
import com.taskbridge.core.models.navigation.NavigationOverlay
import com.taskbridge.core.models.navigation.NavigationPath
import kotlinx.coroutines.flow.Flow

class NavigationRepositoryImpl(
    private val interactor: NavigationInteractor
) : NavigationRepository {
    override val activePath: Flow<NavigationPath?> = interactor.activePath
    override val currentTab: Flow<AppTab> = interactor.currentTab
    override val overlay: Flow<NavigationOverlay?> = interactor.overlay

    override suspend fun selectTab(tab: AppTab) {
        interactor.selectTab(tab)
    }

    override suspend fun pushDestination(destination: NavigationDestination) {
        interactor.pushDestination(destination)
    }

    override suspend fun popDestination() {
        interactor.popDestination()
    }

    override suspend fun pullToRoot(tab: AppTab) {
        interactor.pullToRoot(tab)
    }

    override suspend fun fetchActivePath(): NavigationPath? {
        return interactor.fetchActivePath()
    }

    override suspend fun fetchCurrentTab(): AppTab {
        return interactor.fetchCurrentTab()
    }

    override suspend fun setNavigationDestinationMessage(message: NavigationDestinationMessage?) {
        interactor.setNavigationDestinationMessage(message)
    }

    override suspend fun consumeNavigationDestinationMessage(scopeId: String): NavigationDestinationMessage? {
        return interactor.consumeNavigationDestinationMessage(scopeId)
    }
}

File: iOS/TaskBridge/Repository/Impl/NavigationRepositoryImpl.swift
import Foundation
import TaskBridgeCore

@MainActor
class NavigationRepositoryImpl: NavigationRepository {
    private let interactor: NavigationInteractor
    
    init(interactor: NavigationInteractor) {
        self.interactor = interactor
    }
    
    var activePath: AsyncStream<NavigationPath?> {
        return createStream(for: interactor.activePath)
    }
    
    var currentTab: AsyncStream<AppTab> {
        return createStream(for: interactor.currentTab)
    }
    
    var overlay: AsyncStream<NavigationOverlay?> {
        return createStream(for: interactor.overlay)
    }
    
    func selectTab(tab: AppTab) async throws {
        try await interactor.selectTab(tab: tab)
    }

    func pushDestination(_ destination: NavigationDestination) async throws {
        try await interactor.pushDestination(destination: destination)
    }

    func popDestination() async throws {
        try await interactor.popDestination()
    }

    func pullToRoot(tab: AppTab) async throws {
        try await interactor.pullToRoot(tab: tab)
    }
}
